File: c++/log.cc
#include <glog/logging.h>

#include "log.h"

using multipaxos::Instance;
using multipaxos::InstanceState::COMMITTED;
using multipaxos::InstanceState::EXECUTED;

bool IsCommitted(multipaxos::Instance const& instance) {
  return instance.state() == multipaxos::InstanceState::COMMITTED;
}
bool IsExecuted(multipaxos::Instance const& instance) {
  return instance.state() == multipaxos::InstanceState::EXECUTED;
}
bool IsInProgress(multipaxos::Instance const& instance) {
  return instance.state() == multipaxos::InstanceState::INPROGRESS;
}

namespace multipaxos {
bool operator==(multipaxos::Command const& a, multipaxos::Command const& b) {
  return a.type() == b.type() && a.key() == b.key() && a.value() == b.value();
}

bool operator==(multipaxos::Instance const& a, multipaxos::Instance const& b) {
  return a.ballot() == b.ballot() && a.index() == b.index() &&
         a.client_id() == b.client_id() && a.state() == b.state() &&
         a.command() == b.command();
}
}  // namespace multipaxos

bool Insert(log_map_t* log, Instance instance) {
  auto i = instance.index();
  auto it = log->find(i);
  if (it == log->end()) {
    (*log)[i] = std::move(instance);
    return true;
  }
  if (IsCommitted(it->second) || IsExecuted(it->second)) {
    CHECK(it->second.command() == instance.command()) << "Insert case2";
    return false;
  }
  if (instance.ballot() > it->second.ballot()) {
    (*log)[i] = std::move(instance);
    return false;
  }
  if (instance.ballot() == it->second.ballot())
    CHECK(it->second.command() == instance.command()) << "Insert case3";
  return false;
}

void Log::Append(Instance instance) {
  std::scoped_lock lock(mu_);

  int64_t i = instance.index();
  if (i <= global_last_executed_)
    return;

  if (Insert(&log_, std::move(instance))) {
    last_index_ = std::max(last_index_, i);
    cv_committable_.notify_all();
  }
}

void Log::Commit(int64_t index) {
  CHECK(index > 0) << "invalid index";

  std::unique_lock lock(mu_);
  auto it = log_.find(index);
  while (it == log_.end()) {
    cv_committable_.wait(lock);
    it = log_.find(index);
  }

  if (IsInProgress(it->second))
    it->second.set_state(COMMITTED);

  if (IsExecutable())
    cv_executable_.notify_one();
}

std::optional<log_result_t> Log::Execute() {
  std::unique_lock lock(mu_);
  while (running_ && !IsExecutable())
    cv_executable_.wait(lock);

  if (!running_)
    return std::nullopt;

  auto it = log_.find(last_executed_ + 1);
  CHECK(it != log_.end());
  Instance* instance = &it->second;
  kvstore::KVResult result = kvstore::Execute(instance->command(), kv_store_);
  instance->set_state(EXECUTED);
  ++last_executed_;
  return {{instance->client_id(), std::move(result)}};
}

void Log::CommitUntil(int64_t leader_last_executed, int64_t ballot) {
  CHECK(leader_last_executed >= 0) << "invalid leader_last_executed";
  CHECK(ballot >= 0) << "invalid ballot";

  std::unique_lock lock(mu_);
  for (auto i = last_executed_ + 1; i <= leader_last_executed; ++i) {
    auto it = log_.find(i);
    if (it == log_.end())
      break;
    CHECK(ballot >= it->second.ballot()) << "CommitUntil case 2";
    if (it->second.ballot() == ballot)
      it->second.set_state(COMMITTED);
  }
  if (IsExecutable())
    cv_executable_.notify_one();
}

void Log::TrimUntil(int64_t leader_global_last_executed) {
  std::scoped_lock lock(mu_);
  while (global_last_executed_ < leader_global_last_executed) {
    ++global_last_executed_;
    auto it = log_.find(global_last_executed_);
    CHECK(it != log_.end() && IsExecuted(it->second)) << "TrimUntil case 1";
    log_.erase(it);
  }
}

log_vector_t Log::Instances() const {
  std::scoped_lock lock(mu_);
  log_vector_t instances;
  for (auto i = global_last_executed_ + 1; i <= last_index_; ++i)
    if (auto it = log_.find(i); it != log_.end())
      instances.push_back(it->second);
  return instances;
}

Instance const* Log::operator[](std::size_t i) const {
  auto it = log_.find(i);
  return it == log_.end() ? nullptr : &it->second;
}

File: go/log/log.go
package log

import (
	"github.com/golang/protobuf/proto"
	"github.com/psu-csl/replicated-store/go/kvstore"
	pb "github.com/psu-csl/replicated-store/go/multipaxos/comm"
	logger "github.com/sirupsen/logrus"
	"sync"
)

func IsCommitted(instance *pb.Instance) bool {
	return instance.GetState() == pb.InstanceState_COMMITTED
}

func IsExecuted(instance *pb.Instance) bool {
	return instance.GetState() == pb.InstanceState_EXECUTED
}

func IsInProgress(instance *pb.Instance) bool {
	return instance.GetState() == pb.InstanceState_INPROGRESS
}

func IsEqualCommand(cmd1, cmd2 *pb.Command) bool {
	return cmd1.GetType() == cmd2.GetType() && cmd1.GetKey() == cmd2.GetKey() &&
		cmd1.GetValue() == cmd2.GetValue()
}

func IsEqualInstance(a, b *pb.Instance) bool {
	return a.GetBallot() == b.GetBallot() && a.GetIndex() == b.GetIndex() &&
		a.GetClientId() == b.GetClientId() && a.GetState() == b.GetState() &&
		IsEqualCommand(a.GetCommand(), b.GetCommand())
}

func Insert(log map[int64]*pb.Instance, instance *pb.Instance) bool {
	i := instance.GetIndex()
	if _, ok := log[i]; !ok {
		log[i] = instance
		return true
	}

	if IsCommitted(log[i]) || IsExecuted(log[i]) {
		if !IsEqualCommand(log[i].GetCommand(), instance.GetCommand()) {
			logger.Panicf("case 2 violation\n")
		}
		return false
	}

	if instance.GetBallot() > log[i].GetBallot() {
		log[i] = instance
		return false
	}

	if instance.GetBallot() == log[i].GetBallot() {
		if !IsEqualCommand(log[i].GetCommand(), instance.GetCommand()) {
			logger.Panicf("case 3 violation\n")
		}
	}
	return false
}

type Log struct {
	running            bool
	kvStore            kvstore.KVStore
	log                map[int64]*pb.Instance
	lastIndex          int64
	lastExecuted       int64
	globalLastExecuted int64
	mu                 sync.Mutex
	cvExecutable       *sync.Cond
	cvCommittable      *sync.Cond
}

func NewLog(s kvstore.KVStore) *Log {
	l := Log{
		running:            true,
		kvStore:            s,
		log:                make(map[int64]*pb.Instance),
		lastIndex:          0,
		lastExecuted:       0,
		globalLastExecuted: 0,
		mu:                 sync.Mutex{},
	}
	l.cvExecutable = sync.NewCond(&l.mu)
	l.cvCommittable = sync.NewCond(&l.mu)
	return &l
}

func (l *Log) LastExecuted() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()

	return l.lastExecuted
}

func (l *Log) GlobalLastExecuted() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()

	return l.globalLastExecuted
}

func (l *Log) AdvanceLastIndex() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()

	l.lastIndex += 1
	return l.lastIndex
}

func (l *Log) SetLastIndex(lastIndex int64) {
	l.mu.Lock()
	defer l.mu.Unlock()

	if lastIndex > l.lastIndex {
		l.lastIndex = lastIndex
	}
}

func (l *Log) LastIndex() int64 {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.lastIndex
}

func (l *Log) Stop() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.running = false
	l.kvStore.Close()
	l.cvExecutable.Signal()
}

func (l *Log) IsExecutable() bool {
	instance, ok := l.log[l.lastExecuted+1]
	if ok && IsCommitted(instance) {
		return true
	}
	return false
}

func (l *Log) Append(instance *pb.Instance) {
	l.mu.Lock()
	defer l.mu.Unlock()

	i := instance.GetIndex()
	if i <= l.globalLastExecuted {
		return
	}

	if Insert(l.log, instance) {
		if i > l.lastIndex {
			l.lastIndex = i
		}
		l.cvCommittable.Broadcast()
	}
}

func (l *Log) Commit(index int64) {
	if index <= 0 {
		logger.Panicf("Index %v < 0\n", index)
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	instance, ok := l.log[index]
	for !ok {
		l.cvCommittable.Wait()
		instance, ok = l.log[index]
	}

	if IsInProgress(instance) {
		instance.State = pb.InstanceState_COMMITTED
	}
	if l.IsExecutable() {
		l.cvExecutable.Signal()
	}
}

func (l *Log) Execute() (int64, *kvstore.KVResult) {
	l.mu.Lock()
	defer l.mu.Unlock()

	for l.running && !l.IsExecutable() {
		l.cvExecutable.Wait()
	}

	if !l.running {
		return -1, nil
	}

	instance, ok := l.log[l.lastExecuted+1]
	if !ok {
		logger.Panicf("Instance at Index %v empty\n", l.lastExecuted+1)
	}
	result := kvstore.Execute(instance.GetCommand(), l.kvStore)
	instance.State = pb.InstanceState_EXECUTED
	l.lastExecuted += 1
	return instance.ClientId, &result
}

func (l *Log) CommitUntil(leaderLastExecuted int64, ballot int64) {
	if leaderLastExecuted < 0 {
		logger.Panic("invalid leader_last_executed in commit_until")
	}
	if ballot < 0 {
		logger.Panic("invalid ballot in commit_until")
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	for i := l.lastExecuted + 1; i <= leaderLastExecuted; i++ {
		instance, ok := l.log[i]
		if !ok {
			break
		}
		if ballot < instance.GetBallot() {
			panic("CommitUntil case 2")
		}
		if instance.GetBallot() == ballot {
			instance.State = pb.InstanceState_COMMITTED
		}
	}
	if l.IsExecutable() {
		l.cvExecutable.Signal()
	}
}

func (l *Log) TrimUntil(leaderGlobalLastExecuted int64) {
	l.mu.Lock()
	defer l.mu.Unlock()

	for l.globalLastExecuted < leaderGlobalLastExecuted {
		l.globalLastExecuted += 1
		instance, ok := l.log[l.globalLastExecuted]
		if !ok || !IsExecuted(instance) {
			logger.Panicln("TrimUntil case 1")
		}
		delete(l.log, l.globalLastExecuted)
	}
}

func (l *Log) Instances() []*pb.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()

	instances := make([]*pb.Instance, 0, len(l.log))
	for i := l.globalLastExecuted + 1; i <= l.lastIndex; i++ {
		instance := proto.Clone(l.log[i]).(*pb.Instance)
		if instance != nil {
			instances = append(instances, instance)
		}
	}
	return instances
}

func (l *Log) At(index int64) *pb.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()
	if instance, ok := l.log[index]; ok {
		return instance
	}
	return nil
}

func (l *Log) GetLog() map[int64]*pb.Instance {
	l.mu.Lock()
	defer l.mu.Unlock()

	logMap := make(map[int64]*pb.Instance)
	for index, instance := range l.log {
		logMap[index] = proto.Clone(instance).(*pb.Instance)
	}
	return logMap
}
